### PR TITLE
[chrome] Add definitions of functions for chrome.storage which support Promise

### PR DESCRIPTION
**Pull request by henev**
This pull request adds definitions for `chrome.storage` functions which return Promise in Manifest V3.

When adding them, I referred the following API reference document:
https://developer.chrome.com/docs/extensions/reference/storage/

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://developer.chrome.com/docs/extensions/reference/storage/

### CODE REVIEW

**Review by matcarlson — COMMENTED** [review]
Would it make sense to you to add something like `@since Chrome 72 - MV3` or `@since Chrome 72; MV2` to each definition, so it's easier for users to see which version they are able to use in their extension?
Note - I just picked a random version, I'm not sure when this was added.

**Comment by henev** [comment]
@matcarlson Not sure where to find that? Any idea?

File: types/chrome/index.d.ts [hunk 1]
```diff
@@ -7184,25 +7184,50 @@ declare namespace chrome.storage {
         getBytesInUse(callback: (bytesInUse: number) => void): void;
```
Comment (matcarlson) [inline types/chrome/index.d.ts:7184]: From what I can tell this definition is useless -- It's technically OK, but will always return 0 in the callback. 
What do you think about removing lines :7179 through :7984? Please LMK if you think I'm off-base, though!

File: types/chrome/index.d.ts [hunk 2]
```diff
@@ -7184,25 +7184,50 @@ declare namespace chrome.storage {
...
          * @param callback Optional.
          * Callback on success, or on failure (in which case runtime.lastError will be set).
          */
         clear(callback?: () => void): void;
+        /**
+         * Sets multiple items.
+         * @param items An object which gives each key/value pair to update storage with. Any other key/value pairs in storage will not be affected.
+         * Primitive values such as numbers will serialize as expected. Values with a typeof "object" and "function" will typically serialize to {}, with the exception of Array (serializes as expected), Date, and Regex (serialize using their String representation).
+         * @return The `set` method provides its result via callback or returned as a `Promise` (MV3 only).
+         */
```
Comment (matcarlson) [inline types/chrome/index.d.ts:7214]: Add `@since MV3` or similar

File: types/chrome/index.d.ts [hunk 3]
```diff
@@ -7184,25 +7184,50 @@ declare namespace chrome.storage {
...
          * Removes all items from storage.
          * @param callback Optional.
          * Callback on success, or on failure (in which case runtime.lastError will be set).
          */
         clear(callback?: () => void): void;
+        /**
+         * Sets multiple items.
+         * @param items An object which gives each key/value pair to update storage with. Any other key/value pairs in storage will not be affected.
+         * Primitive values such as numbers will serialize as expected. Values with a typeof "object" and "function" will typically serialize to {}, with the exception of Array (serializes as expected), Date, and Regex (serialize using their String representation).
+         * @return The `set` method provides its result via callback or returned as a `Promise` (MV3 only).
```
Comment (matcarlson) [inline types/chrome/index.d.ts:7213]: I'd not mention the callback here, since this definition doesn't have it. You could shorten this to `@return A void Promise`, but it's up to you.

File: types/chrome/index.d.ts [hunk 4]
```diff
@@ -7184,25 +7184,50 @@ declare namespace chrome.storage {
...
+         */
+        set(items: { [key: string]: any }): Promise<void>;
         /**
          * Sets multiple items.
          * @param items An object which gives each key/value pair to update storage with. Any other key/value pairs in storage will not be affected.
          * Primitive values such as numbers will serialize as expected. Values with a typeof "object" and "function" will typically serialize to {}, with the exception of Array (serializes as expected), Date, and Regex (serialize using their String representation).
          * @param callback Optional.
          * Callback on success, or on failure (in which case runtime.lastError will be set).
```
Comment (matcarlson) [inline types/chrome/index.d.ts:7221]: Optional - Can you add `@since MV2` after this line?

File: types/chrome/index.d.ts [hunk 5]
```diff
@@ -7184,25 +7184,50 @@ declare namespace chrome.storage {
...
          */
-        set(items: Object, callback?: () => void): void;
+        set(items: { [key: string]: any }, callback?: () => void): void;
+        /**
+         * Removes one or more items from storage.
+         * @param keys A single key or a list of keys for items to remove.
+         * @param callback Optional.
+         * @return The `remove` method provides its result via callback or returned as a `Promise` (MV3 only).
+         */
```
Comment (matcarlson) [inline types/chrome/index.d.ts:7229]: Can you add `@since MV3` or `@since MV3 only` here?

File: types/chrome/index.d.ts [hunk 6]
```diff
@@ -7184,25 +7184,50 @@ declare namespace chrome.storage {
...
          * @param callback Optional.
          * Callback on success, or on failure (in which case runtime.lastError will be set).
          */
-        set(items: Object, callback?: () => void): void;
+        set(items: { [key: string]: any }, callback?: () => void): void;
+        /**
+         * Removes one or more items from storage.
+         * @param keys A single key or a list of keys for items to remove.
+         * @param callback Optional.
+         * @return The `remove` method provides its result via callback or returned as a `Promise` (MV3 only).
```
Comment (matcarlson) [inline types/chrome/index.d.ts:7228]: Similar to the prior comments, I don't think it's useful to mention the callback here since it's not included on line 7230. I'd say `@return A void Promise`.

File: types/chrome/index.d.ts [hunk 7]
```diff
@@ -7211,19 +7236,26 @@ declare namespace chrome.storage {
...
          * @param callback Callback with storage items, or on failure (in which case runtime.lastError will be set).
          * Parameter items: Object with items in their key-value mappings.
          */
         get(callback: (items: { [key: string]: any }) => void): void;
+        /**
+         * Gets one or more items from storage.
+         * @param keys A single key to get, list of keys to get, or a dictionary specifying default values.
+         * An empty list or object will return an empty result object. Pass in null to get the entire contents of storage.
+         * @return The `get` method provides its result via callback or returned as a `Promise` (MV3 only).
+         */
```
Comment (matcarlson) [inline types/chrome/index.d.ts:7249]: Could you add `@since MV3` or `@since MV3 only` here as well?

File: types/chrome/index.d.ts [hunk 8]
```diff
@@ -7211,19 +7236,26 @@ declare namespace chrome.storage {
...
+        get(keys?: string | string[] | { [key: string]: any } | null): Promise<{ [key: string]: any }>;
         /**
          * Gets one or more items from storage.
          * @param keys A single key to get, list of keys to get, or a dictionary specifying default values.
          * An empty list or object will return an empty result object. Pass in null to get the entire contents of storage.
          * @param callback Callback with storage items, or on failure (in which case runtime.lastError will be set).
          * Parameter items: Object with items in their key-value mappings.
```
Comment (matcarlson) [inline types/chrome/index.d.ts:7256]: Optional - Could you add `@since MV2`?

File: types/chrome/index.d.ts [hunk 9]
```diff
@@ -7184,25 +7184,50 @@ declare namespace chrome.storage {
...
+        /**
+         * Gets the amount of space (in bytes) being used by one or more items.
+         * @param keys Optional. A single key or list of keys to get the total usage for. An empty list will return 0. Pass in null to get the total usage of all of storage.
          * @param callback Callback with the amount of space being used by storage, or on failure (in which case runtime.lastError will be set).
          * Parameter bytesInUse: Amount of space being used in storage, in bytes.
          */
         getBytesInUse(keys: string | string[] | null, callback: (bytesInUse: number) => void): void;
+        /**
+         * Removes all items from storage.
+         * @return The `clear` method provides its result via callback or returned as a `Promise` (MV3 only).
```
Comment (matcarlson) [inline types/chrome/index.d.ts:7200]: I'd add `@since MV3` here as well.

**Review by matcarlson — COMMENTED** [review]
I'd actually just mark at least the Manifest v3 ones as `@since MV3` or `@since MV3 only`. After your response I took a look at the docs and saw that they weren't listed. 

Here's a more-proper review

File: types/chrome/index.d.ts [hunk 10]
```diff
@@ -7184,25 +7184,50 @@ declare namespace chrome.storage {
...
-         * @param keys A single key or list of keys to get the total usage for. An empty list will return 0. Pass in null to get the total usage of all of storage.
+         * @param keys Optional. A single key or list of keys to get the total usage for. An empty list will return 0. Pass in null to get the total usage of all of storage.
+         * @return The `getBytesInUse` method provides its result via callback or returned as a `Promise` (MV3 only).
+         */
+        getBytesInUse(keys?: string | string[] | null): Promise<number>;
+        /**
+         * Gets the amount of space (in bytes) being used by one or more items.
+         * @param keys Optional. A single key or list of keys to get the total usage for. An empty list will return 0. Pass in null to get the total usage of all of storage.
          * @param callback Callback with the amount of space being used by storage, or on failure (in which case runtime.lastError will be set).
          * Parameter bytesInUse: Amount of space being used in storage, in bytes.
```
[the remark this answers is not in the document]
Comment (hristo-enev) [reply types/chrome/index.d.ts:7195]: I don't think adding `@since MV2` is needed.

File: types/chrome/index.d.ts (hunk 1 above)
Comment (henev) [reply types/chrome/index.d.ts:7184]: The problem here if this is removed is the following:

`chrome.storage.sync.getBytesInUse(() => {});` which should be valid gives error because it is taking the first type which returns promise and accepts keys as an argument.

File: types/chrome/index.d.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (henev) [reply types/chrome/index.d.ts:7195]: Don't think this is necessary. I don't even know if this is true.

File: types/chrome/index.d.ts (hunk 1 above)
Comment (matcarlson) [reply types/chrome/index.d.ts:7184]: Ah - I didn't think this would cause an error with just `.d.ts` files. Thanks, and sorry for sending you astray!

**Review by matcarlson — APPROVED** [review]
Thanks so much for making the definitions more accurate, and for testing the changes I mentioned. LGTM

**Comment by henev** [comment]
Ready to merge
